ethicalgardeners/metricscollector.py

```diff
@@ -0,0 +1,320 @@
+        metrics are tracked correctly for each new simulation run.
+        """
+        self.metrics = {
+            "step": 0,
+            "total_planted_flowers": 0,
+            "num_planted_flowers_per_agent": {},
+            "total_harvested_flowers": 0,
+            "num_harvested_flowers_per_agent": {},
+            "avg_pollution_percent": 0.0,
+            "num_cells_pollution_above_90": 0,
+            "num_cells_pollution_above_75": 0,
+            "num_cells_pollution_above_50": 0,
+            "num_cells_pollution_above_25": 0,
+            "rewards": {},
+            "accumulated_rewards": {},
+            "agent_selection": None,
+        }
+
+    def export_metrics(self):
+        """
+        Export collected metrics to a local file.
+
+        This method exports the current metrics to a CSV file in the specified
+        output directory if export_on is True. A new file is created for each
+        run of the program, and metrics are appended to this file at each call.
+        """
+        if self.export_on:
+            try:
+                import csv
+            except ImportError:
+                raise ImportError(
+                    "Error while importing csv module. "
+                )
+            try:
+                import os
+            except ImportError:
+                raise ImportError(
+                    "Error while importing os module. "
+                )
+
+            # Create output directory if it doesn't exist
+            if not os.path.exists(self.out_dir_path):
+                os.makedirs(self.out_dir_path)
+
+            # Generate filename with run_id
+            filename = os.path.join(self.out_dir_path,
+                                    f"metrics_run_{self._run_id}.csv")
+
+            # Prepare row with metrics
+            metrics_row = self._prepare_metrics()
+
+            # Check if file exists to determine if we need to write headers
+            file_exists = os.path.isfile(filename)
+
+            # Write or append to CSV
+            with open(filename, 'a' if file_exists else 'w',
+                      newline='') as csvfile:
+                writer = csv.DictWriter(csvfile,
+                                        fieldnames=list(metrics_row.keys()))
+
+                if not file_exists:
+                    writer.writeheader()
+
+                writer.writerow(metrics_row)
+
+    def send_metrics(self):
+        """
+        Send collected metrics to Weights and Biases (WandB).
+
+        This method sends the current metrics to WandB for visualization and
+        experiment tracking if send_on is True. It automatically handles WandB
+        initialization if needed and logs all relevant metrics from the current
+        simulation state.
+        """
+        if self.send_on:
+            try:
+                import wandb
+            except ImportError:
+                raise ImportError(
+                    "Error while importing wandb module. "
```

Comment:
Here, we can leave this `try/except` because WandB is not standard. However, the message error as it stands is not really useful: the basic ImportError already tells us that the `wandb` module is missing!

What we can do is to tell the user an appropriate message for *our* context. For example, we might place the `wandb` requirement as part of an "extra" list, and tell them "WandB is required to use `send_metrics`. Please install WandB with `pip install wandb` or `pip install ethicalgardeners[metrics]`"